File: Src/LoLMatchAccepterNet/LCU/Game.cs
using System.Text.Json;

namespace LoLMatchAccepterNet.LCU
{
    public class Game
    {
        public const string InProgress = "InProgress";
        public const string GameStart = "GameStart";
        public const string LoadingScreen = "LoadingScreen";
        public const string None = "None";
        public const string Lobby = "Lobby";
        public const string ReadyCheck = "ReadyCheck";
        public const string Matchmaking = "Matchmaking";
        public const string ChampSelect = "ChampSelect";

        private readonly HttpClient _client;
        private readonly string _baseUrl;

        public Game(HttpClient client, int port)
        {
            _client = client;
            this._baseUrl = $"https://127.0.0.1:{port}";
        }

        public async Task<bool> IsActive()
        {
            try
            {
                string gamePhase = await GetGamePhase();

                if (gamePhase == InProgress ||
                    gamePhase == GameStart || gamePhase == LoadingScreen)
                {
                    return true;
                }

                var spectatorResponse = await _client.GetAsync($"{_baseUrl}/lol-spectator/v1/spectate/active-games/for-summoner/0");
                if (spectatorResponse.StatusCode == System.Net.HttpStatusCode.OK)
                {
                    // Status 200 means there's an active game
                    return true;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error checking game status: {ex.Message}");
            }

            return false;
        }

        public async Task WaitUntilGameEnds()
        {
            bool gameEnded = false;

            while (!gameEnded)
            {
                try
                {

                    string gamePhase = await GetGamePhase();

                    //Game not found
                    if (string.IsNullOrEmpty(gamePhase))
                    {
                        return;
                    }

                    if (gamePhase == "None" || gamePhase == "Lobby" || gamePhase == "Matchmaking")
                    {
                        gameEnded = true;
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error checking if game ended: {ex.Message}");
                    return;
                }

                await Task.Delay(1000);
            }
        }

        public async Task<string> WaitUntilPhaseEnds(params string[] gamePhases)
        {
            while (true)
            {
                try
                {
                    string currentPhase = await GetGamePhase();
                    if (!gamePhases.Contains(currentPhase))
                    {
                        return currentPhase;
                    }
                    await Task.Delay(1000);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error checking game phase: {ex.Message}");
                    return string.Empty;
                }
            }
        }

        public async Task<string> GetGamePhase()
        {
            var gameSessionResponse = await _client.GetAsync($"{_baseUrl}/lol-gameflow/v1/session");
            if (gameSessionResponse.IsSuccessStatusCode)
            {
                string content = await gameSessionResponse.Content.ReadAsStringAsync();
                JsonDocument doc = JsonDocument.Parse(content);

                if (doc.RootElement.TryGetProperty("phase", out JsonElement phase))
                {
                    string gamePhase = phase.GetString()!;

                    return gamePhase;
                }
            }
            return string.Empty;
        }

        public async Task<bool> WaitForQueue()
        {
            var matchResponse = await _client.GetAsync($"{_baseUrl}/lol-matchmaking/v1/search");
            if (matchResponse.IsSuccessStatusCode)
            {
                string content = await matchResponse.Content.ReadAsStringAsync();
                JsonDocument doc = JsonDocument.Parse(content);

                doc.RootElement.TryGetProperty("searchState", out JsonElement searchState);
                if (searchState.GetString() == "Found")
                {
                    Console.WriteLine("Match found! Accepting...");
                    var acceptResponse = await _client.PostAsync(
                        $"{_baseUrl}/lol-matchmaking/v1/ready-check/accept",
                        null
                    );

                    if (acceptResponse.IsSuccessStatusCode)
                    {
                        Console.WriteLine("Match accepted successfully!");
                        return true;
                    }
                }
            }
            return false;
        }
    }
}

File: Src/LoLMatchAccepterCpp/src/LCU/Game.cpp
#include "Game.h"
#include "../json.hpp"
#include <iostream>
#include <thread>
#include <chrono>
#include <vector>
#include <algorithm>

using json = nlohmann::json;

Game::Game(std::shared_ptr<LcuHttpClient> client) : _client(client) {
}

bool Game::IsActive() {
    try {
        std::string gamePhase = GetGamePhase();

        if (gamePhase == InProgress || gamePhase == GameStart || gamePhase == LoadingScreen) {
            return true;
        }

        HttpResponse spectatorResponse = _client->Get("/lol-spectator/v1/spectate/active-games/for-summoner/0");
        if (spectatorResponse.IsSuccess()) {
            return true;
        }
    }
    catch (...) {
    }

    return false;
}

void Game::WaitUntilGameEnds() {
    bool gameEnded = false;

    while (!gameEnded) {
        try {
            std::string gamePhase = GetGamePhase();

            if (gamePhase.empty()) {
                return;
            }

            if (gamePhase == None || gamePhase == Lobby || 
                gamePhase == Matchmaking || gamePhase == EndOfGame || 
                gamePhase == WaitingForStats) {
                gameEnded = true;
            }
        }
        catch (const std::exception& ex) {
            std::cout << "Error checking if game ended: " << ex.what() << std::endl;
            return;
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    }
}

std::string Game::WaitUntilPhaseEnds(const std::vector<std::string>& gamePhases) {
    while (true) {
        try {
            std::string currentPhase = GetGamePhase();
            
            if (std::find(gamePhases.begin(), gamePhases.end(), currentPhase) == gamePhases.end()) {
                return currentPhase;
            }
            
            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
        }
        catch (const std::exception& ex) {
            std::cout << "Error checking game phase: " << ex.what() << std::endl;
            return "";
        }
    }
}

std::string Game::GetGamePhase() {
    HttpResponse response = _client->Get("/lol-gameflow/v1/session");

    if (!response.IsSuccess() || response.body.empty()) {
        return "";
    }

    try {
        auto jsonResponse = json::parse(response.body);
        if (jsonResponse.contains("phase")) {
            return jsonResponse["phase"].get<std::string>();
        }
    }
    catch (...) {
    }

    return "";
}

void Game::StartQueue() {
    try {
        // Check if we're in the lobby
        std::string currentPhase = GetGamePhase();
        if (currentPhase != Lobby) {
            return;
        }

        // Get current lobby information
        HttpResponse lobbyResponse = _client->Get("/lol-lobby/v2/lobby");

        if (!lobbyResponse.IsSuccess() || lobbyResponse.body.empty()) {
            return;
        }

        auto lobbyData = json::parse(lobbyResponse.body);

        // Check if we are the lobby owner
        if (lobbyData.contains("localMember") && lobbyData.contains("members")) {
            bool isOwner = false;

            if (lobbyData["localMember"].contains("isLeader")) {
                isOwner = lobbyData["localMember"]["isLeader"].get<bool>();
            }

            if (isOwner) {
                std::cout << "Starting queue as lobby owner..." << std::endl;
                HttpResponse startQueueResult = _client->Post("/lol-lobby/v2/lobby/matchmaking/search");

                if (startQueueResult.IsSuccess()) {
                    std::cout << "Queue started successfully!" << std::endl;
                } else {
                    std::cout << "Failed to start queue (Status: " << startQueueResult.statusCode << ")" << std::endl;
                    std::this_thread::sleep_for(std::chrono::milliseconds(4000));
                }
            }
        }
    }
    catch (const std::exception& ex) {
        std::cout << "Error starting queue: " << ex.what() << std::endl;
    }
}

bool Game::WaitForQueue() {
    HttpResponse matchResponse = _client->Get("/lol-matchmaking/v1/search");

    if (matchResponse.IsSuccess() && !matchResponse.body.empty()) {
        try {
            auto jsonResponse = json::parse(matchResponse.body);
            if (jsonResponse.contains("searchState")) {
                std::string searchState = jsonResponse["searchState"].get<std::string>();
                if (searchState == "Found") {
                    std::cout << "Match found! Accepting..." << std::endl;

                    HttpResponse acceptResult = _client->Post("/lol-matchmaking/v1/ready-check/accept");

                    if (acceptResult.IsSuccess()) {
                        std::cout << "Match accepted successfully!" << std::endl;
                        return true;
                    }
                }
            }
        }
        catch (...) {
        }
    }

    std::string gamePhase = GetGamePhase();
    return gamePhase == ChampSelect || gamePhase == InProgress;
}

void Game::NavigateToLobby() {
    try {
        std::cout << "Navigating to lobby..." << std::endl;

        WaitUntilPhaseEnds({WaitingForStats});

        std::cout << "Skipping honor screen..." << std::endl;
        HttpResponse skipHonorResult = _client->Post("/lol-honor-v2/v1/honor-player");

        if (skipHonorResult.IsSuccess()) {
            std::cout << "Honor screen skipped" << std::endl;
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(500));

        HttpResponse playAgainResult = _client->Post("/lol-lobby/v2/play-again");

        if (playAgainResult.IsSuccess()) {
            std::cout << "Successfully navigated to lobby!" << std::endl;
            return;
        }
    }
    catch (const std::exception& ex) {
        std::cout << "Error navigating to lobby: " << ex.what() << std::endl;
    }
}
